File: test23/list.cpp
#include "list.h"

List *createList()
{
    List *list = new List{nullptr,nullptr};
    return list;
}

bool isEmpty(List *list)
{
    return list->head == nullptr && list->tail == nullptr;
}

void add(List *list, int value)
{
    if (isEmpty(list))
    {
        list->head = new ListElement;
        list->head->value = value;
        list->head->next = nullptr;
        list->tail = list->head;
    }
    else
    {
        ListElement *tmp = new ListElement;
        tmp->value = value;
        tmp->next = nullptr;
        list->tail->next = tmp;
        list->tail = list->tail->next;
    }
}

void change(List *list, int index, int value)
{
    ListElement *tmp = list->head;
    while (index > 0 && tmp->next != nullptr)
    {
        tmp = tmp->next;
        index--;
    }

    tmp->value = value;
}

void swapElements(List *list, int index, int position)
{
    ListElement *tmp = list->head;
    while (index > 0 && tmp->next != nullptr)
    {
        tmp = tmp->next;
        index--;
    }

    ListElement *temp = list->head;
    while (position > 0 && tmp->next != nullptr)
    {
        temp = temp->next;
        position--;
    }

    int t = tmp->value;
    tmp->value = temp->value;
    temp->value = t;
}

int get(List *list, int index)
{
    if (isEmpty(list))
        return -42;

    ListElement *tmp = list->head;
    while (tmp->next != nullptr && index > 0)
    {
        tmp = tmp->next;
        index--;
    }

    return tmp->value;
}

void deleteList(List *list)
{
    while(!isEmpty(list))
    {
        if (list->head == list->tail)
        {
            delete list->head;
            list->head = nullptr;
            list->tail = nullptr;

            return;
        }

        ListElement *tmp = list->head;
        list->head = list->head->next;
        delete tmp;
    }
}

File: java.1/2/List.java
package com.spbu.group144.Dementiev;

public class List {
    private class ListNode {
        private ListNode next;
        private int value;

        public ListNode (int value) {
            this.next = null;
            this.value = value;
        }
    }

    private ListNode head = null;
    private ListNode tail = null;
    private int size;
    List() {
        size = 0;
    }

    public int getSize() {
        return size;
    }

    private boolean isEmpty() {
        return (size == 0);
    }

    public int peekBegin() {
        if (isEmpty())
            return 900009;
        return head.value;
    }

    public void push(int value) {
        ListNode newNode = new ListNode(value);
        newNode.next = head;
        head = newNode;
        size++;
    }

    public void push(int value, int index) {
        ListNode tmp = head;
        for (int i = 0; i < index - 1; i++)
            tmp = tmp.next;

        ListNode newNode = new ListNode(value);
        newNode.next = tmp.next;
        tmp.next = newNode;
        size++;
    }

    public int pop() {
        if (isEmpty())
            return -1;

        int data = head.value;
        head = head.next;
        size--;
        return data;
    }

    public int pop(int index) {
        if (index == 0) {
            return pop();
        }
        if (index > getSize()) {
            return -1;
        }
        ListNode tmp = head;
        for (int i = 0; i < index - 2; i++)
            tmp = tmp.next;
        int result = tmp.next.value;
        tmp.next = tmp.next.next;
        size--;
        return result;
    }

    public int find(int value) {
        ListNode tmp = head;
        for (int i = 0; i < getSize(); i++) {
            if (tmp.value == value)
                return i;
            tmp = tmp.next;
        }
        return -1;
    }
}
